src/java.base/share/classes/java/net/HostPortrange.java

```diff
@@ -186,7 +186,8 @@ static String toLowerCase(String s) {
                 }
                 sb.append((char)(c - CASE_DIFF));
             } else {
-                throw new IllegalArgumentException("Invalid characters in hostname");
+                final String invalidChar = String.format("\\u%04x", (int) c);
```

Comment:
Why not construct the whole message using String.format? It would avoid the string concatenation below...

Reply:
That's certainly better. I've now updated the PR to follow that suggestion. The new test continues to pass.